### PR TITLE
test(esbuild): migrate tests cases of packagejson

### LINKED ISSUE #783: Migrate esbuild's tests of `packagejson`
We want to migrate esbuild's tests defined in https://github.com/evanw/esbuild/blob/main/internal/bundler_tests/bundler_packagejson_test.go to this repo.

## Guidance

Take the first test [`TestPackageJsonMain`](https://github.com/evanw/esbuild/blob/44e746965d783646f97daf3d0617ff816727e7fb/internal/bundler_tests/bundler_packagejson_test.go#L13-L37) as an example.

1. All you do is creating corresponding files like I did in https://github.com/rolldown/rolldown/pull/782.

2. After create the test files, try pass the test. __However, it doesn't matter if the test is failed. If the test is failed, just rename the test folder from `main` to `.main`. And it would be skipped automatically__

3. Remove the test name you just added in this test list https://github.com/rolldown/rolldown/pull/781/files

3. Commit the change with message like `test(esbuild/packagejson): add ....`. Remember run `just roll` to see if everything is ok.

4. Open a PR.

### PR DESCRIPTION
**Pull request by zhusjfaker**
### Description

base the https://github.com/rolldown/rolldown/issues/783 issue link 

i have migrate  78 case in esbuild main branch for rolldown esbuild packagejson fold 

but the process has not been finished 

maybe i will finish it at the end of the April

### CODE REVIEW

**Comment by zhusjfaker** [comment]
Fixes #783

File: crates/rolldown/tests/esbuild/packagejson/TestPackageJsonBadMain/node_modules/demo-pkg/package.json
Comment (hyf0) [inline crates/rolldown/tests/esbuild/packagejson/TestPackageJsonBadMain/node_modules/demo-pkg/package.json:1]: Please use  snake_case here. Change `TestPackageJsonBadMain` to `bad_main`.

**Review by hyf0 — CHANGES_REQUESTED** [review]

**Comment by hyf0** [comment]
All looks well. Well done!

**Comment by hyf0** [comment]
> but the process has not been finished

I suggest you could merge the current added tests. However, it's also fine that if you want merge them all together in a single PR.

If you are ready to merge the PR, just undraft this PR and I'll starting to review or and merge it.

File: crates/rolldown/tests/esbuild/packagejson/TestPackageJsonBadMain/node_modules/demo-pkg/package.json
Comment (zhusjfaker) [reply crates/rolldown/tests/esbuild/packagejson/TestPackageJsonBadMain/node_modules/demo-pkg/package.json:1]: ok

**Comment by zhusjfaker** [comment]
> > but the process has not been finished
> 
> I suggest you could merge the current added tests. However, it's also fine that if you want merge them all together in a single PR.
> 
> If you are ready to merge the PR, just undraft this PR and I'll starting to review or and merge it.

ok

**Comment by zhusjfaker** [comment]
```golang

func TestPackageJsonNodePathsIssue2752(t *testing.T) {
	packagejson_suite.expectBundled(t, bundled{
		files: map[string]string{
			"/src/entry.js": `
				import "pkg1"
				import "pkg2"
				import "@scope/pkg3/baz"
				import "@scope/pkg4"
			`,
			"/usr/lib/pkg/pkg1/package.json":          `{ "main": "./foo.js" }`,
			"/usr/lib/pkg/pkg1/foo.js":                `console.log('pkg1')`,
			"/lib/pkg/pkg2/package.json":              `{ "exports": { ".": "./bar.js" } }`,
			"/lib/pkg/pkg2/bar.js":                    `console.log('pkg2')`,
			"/var/lib/pkg/@scope/pkg3/package.json":   `{ "browser": { "./baz.js": "./baz-browser.js" } }`,
			"/var/lib/pkg/@scope/pkg3/baz-browser.js": `console.log('pkg3')`,
			"/tmp/pkg/@scope/pkg4/package.json":       `{ "exports": { ".": { "import": "./bat.js" } } }`,
			"/tmp/pkg/@scope/pkg4/bat.js":             `console.log('pkg4')`,
		},
		entryPaths: []string{"/src/entry.js"},
		options: config.Options{
			Mode:          config.ModeBundle,
			AbsOutputFile: "/out.js",
			AbsNodePaths: []string{
				"/usr/lib/pkg",
				"/lib/pkg",
				"/var/lib/pkg",
				"/tmp/pkg",
			},
		},
	})
}
```

This case simulates a problem that occurs with a global module lookup, and I can't reproduce it too well in the current hard disk project, so I didn't migrate this case. 

@hyf0

**Comment by zhusjfaker** [comment]
Except for TestPackageJsonNodePathsIssue2752 I've migrated all the other files needed for the test cases, specifically how to set up the '_config.json' against the product of esbuild How to go about combining the rolldown comparisons, including reporting errors. I still need to discuss with you.   @hyf0

**Comment by hyf0** [comment]
> Except for TestPackageJsonNodePathsIssue2752 I've migrated all the other files needed for the test cases, specifically how to set up the '_config.json' against the product of esbuild How to go about combining the rolldown comparisons, including reporting errors. I still need to discuss with you. @hyf0

Just follow how #782 did.

For `_config.json`, you could references:
- https://rolldown.rs/contrib-guide/testing#how-test-cases-are-defined
- Other existed `_config.json`

For non-passed cases, there's need to discuss them in this PR, just rename them as I said in

> However, it doesn't matter if the test is failed. If the test is failed, just rename the test folder from main to .main. And it would be skipped automatically https://github.com/rolldown/rolldown/issues/783

Run `just test-rust` and [update snapshots](https://rolldown.rs/contrib-guide/testing#snapshot-testing). If some tests are failed, just ignore them by renaming.

**Comment by hyf0** [comment]
> ```go
> func TestPackageJsonNodePathsIssue2752(t *testing.T) {
> 	packagejson_suite.expectBundled(t, bundled{
> 		files: map[string]string{
> 			"/src/entry.js": `
> 				import "pkg1"
> 				import "pkg2"
> 				import "@scope/pkg3/baz"
> 				import "@scope/pkg4"
> 			`,
> 			"/usr/lib/pkg/pkg1/package.json":          `{ "main": "./foo.js" }`,
> 			"/usr/lib/pkg/pkg1/foo.js":                `console.log('pkg1')`,
> 			"/lib/pkg/pkg2/package.json":              `{ "exports": { ".": "./bar.js" } }`,
> 			"/lib/pkg/pkg2/bar.js":                    `console.log('pkg2')`,
> 			"/var/lib/pkg/@scope/pkg3/package.json":   `{ "browser": { "./baz.js": "./baz-browser.js" } }`,
> 			"/var/lib/pkg/@scope/pkg3/baz-browser.js": `console.log('pkg3')`,
> 			"/tmp/pkg/@scope/pkg4/package.json":       `{ "exports": { ".": { "import": "./bat.js" } } }`,
> 			"/tmp/pkg/@scope/pkg4/bat.js":             `console.log('pkg4')`,
> 		},
> 		entryPaths: []string{"/src/entry.js"},
> 		options: config.Options{
> 			Mode:          config.ModeBundle,
> 			AbsOutputFile: "/out.js",
> 			AbsNodePaths: []string{
> 				"/usr/lib/pkg",
> 				"/lib/pkg",
> 				"/var/lib/pkg",
> 				"/tmp/pkg",
> 			},
> 		},
> 	})
> }
> ```
> 
> This case simulates a problem that occurs with a global module lookup, and I can't reproduce it too well in the current hard disk project, so I didn't migrate this case.
> 
> @hyf0

That's fine. Make sure you update `https://github.com/rolldown/rolldown/blob/main/crates/rolldown/tests/esbuild/packagejson/tests.txt` correctly.

File: crates/rolldown/tests/esbuild/packagejson/TestPackageJsonBadMain/node_modules/demo-pkg/package.json
Comment (hyf0) [reply crates/rolldown/tests/esbuild/packagejson/TestPackageJsonBadMain/node_modules/demo-pkg/package.json:1]: Following the naming pattern

```
TestPackageJsonBadMain => `bad_main`
TestPackageJsonSyntaxErrorComment => `synxtax_error_comment`
TestPackageJsonSyntaxErrorTrailingComma => `synxtax_error_trailing_comma`
```

File: crates/rolldown/tests/esbuild/packagejson/test_package_json_bad_main/src/entry.js
Comment (hyf0) [inline crates/rolldown/tests/esbuild/packagejson/test_package_json_bad_main/src/entry.js:1]: Following the naming pattern

```
TestPackageJsonBadMain => `bad_main`
TestPackageJsonSyntaxErrorComment => `synxtax_error_comment`
TestPackageJsonSyntaxErrorTrailingComma => `synxtax_error_trailing_comma`
```

`crates/rolldown/tests/esbuild/packagejson/test_package_json_bad_main` should be `crates/rolldown/tests/esbuild/packagejson/bad_main`

File: crates/rolldown/tests/esbuild/packagejson/TestPackageJsonBadMain/node_modules/demo-pkg/package.json
Comment (hyf0) [reply crates/rolldown/tests/esbuild/packagejson/TestPackageJsonBadMain/node_modules/demo-pkg/package.json:1]: Well. I think again. The current naming of this PR is fine due there is a case that `TestCommonJSVariableInESMTypeModule` doesn't have the prefix `TestPackageJson`. So just ignore this comment.

File: crates/rolldown/tests/esbuild/packagejson/test_package_json_bad_main/src/entry.js
Comment (hyf0) [reply crates/rolldown/tests/esbuild/packagejson/test_package_json_bad_main/src/entry.js:1]: Ignore this comment https://github.com/rolldown/rolldown/pull/969#discussion_r1579819015

File: crates/rolldown/tests/esbuild/packagejson/TestPackageJsonBadMain/node_modules/demo-pkg/package.json
Comment (zhusjfaker) [reply crates/rolldown/tests/esbuild/packagejson/TestPackageJsonBadMain/node_modules/demo-pkg/package.json:1]: > Well. I think again. The current naming of this PR is fine due there is a case that `TestCommonJSVariableInESMTypeModule` doesn't have the prefix `TestPackageJson`. So just ignore this comment.

ok

**Comment by hyf0** [comment]
If you feel it's difficult to write testing configs, I could take the PR from here and write those config. That's fine.

**Comment by zhusjfaker** [comment]
I've got all the _config.json ready to go but found some issues with the Resolver for the current rolldown, do I need to fix it. I have prefixed all the problematic folders with “.” added the modifier, but the test won't skip it and still reports an error?  

@hyf0

**Comment by zhusjfaker** [comment]
And now the judgment exception is just a simple boolean, is it necessary to add the exact error message to determine whether the error, and the capture of the same, in addition to build the content of the result of the assertion, whether it is also possible to configure through the config.json to go?

File: crates/rolldown/tests/esbuild/packagejson/tests.txt
Comment (hyf0) [inline crates/rolldown/tests/esbuild/packagejson/tests.txt:1]: Could you delete tests you have added in this PR?

Comment (zhusjfaker) [reply crates/rolldown/tests/esbuild/packagejson/tests.txt:1]: ok

Comment (zhusjfaker) [reply crates/rolldown/tests/esbuild/packagejson/tests.txt:1]: i have deleted  the case which i have add it in rolldown test folder

**Review by hyf0 — APPROVED** [review]
Great works! Thank you!

**Pull request merged**